### ISSUE TITLE
Set Up Experiment Tracking (Weights & Biases)

### ISSUE THREAD

**Issue by sachelsout** [sachelsout/AirSketch#3]
**Depends on:** #1 (repo scaffolded), #2 (Zaratan environment configured)
**Blocks:** #10 (training loop), #11 (hyperparameter sweep)

---

### Background

Without experiment tracking, hyperparameter sweeps become a spreadsheet nightmare — someone updates a number in a Slack message, another person reruns the wrong config, and by the time results are due nobody agrees on which checkpoint is the best one. This issue wires up Weights & Biases (W&B) before any training code is written so that every run from issue #10 onward is automatically logged, versioned, and comparable.

W&B is chosen over MLflow for this project for three reasons: it requires zero infrastructure to host (MLflow needs a tracking server), it has a free tier that covers all student usage, and its sweep agent in issue #11 replaces writing custom grid search code entirely.

Complete this issue before the training loop in issue #10. The training loop should call into the logging interface defined here, not the other way around.

---

### 1. Account and Project Setup

**Step 1 — Create a shared W&B team**

One team member creates the W&B account and project. Everyone else joins it.

- Go to [[wandb.ai](https://wandb.ai/)](https://wandb.ai) and sign up with your UMD email
- Create a new **team** named `airsketch-umd` (free for academic use)
- Create a new **project** inside the team named `airsketch`
- Invite all four team members to the team via Settings → Members
- Copy the team's API key from Settings → API Keys — every member needs this

**Step 2 — Authenticate on every machine (local + Zaratan)**

```bash
# Install wandb if not already in requirements.txt
pip install wandb==0.17.4

# Log in — paste the API key when prompted
wandb login

# Verify authentication worked
wandb whoami
# Expected output: your username and the airsketch-umd team
```

On Zaratan, run `wandb login` once after activating the conda environment. The credentials are stored in `~/.netrc` and persist across sessions — you don't need to log in again per job.

If the cluster has no outbound internet (some HPCs block it), add this to your SLURM scripts:

```bash
# Run in offline mode during the job, sync afterward
export WANDB_MODE=offline

# After the job completes, sync from the login node
wandb sync logs/wandb/run-<timestamp>-<id>/
```

---

### 2. Logging Interface

Rather than scattering `wandb.log()` calls throughout the codebase, create a thin wrapper in `src/logger.py`. This means if the team ever needs to swap W&B for something else, there is exactly one file to change.

**`src/logger.py`**

```python
"""
Experiment tracking wrapper for AirSketch.
Wraps Weights & Biases. All training code calls this module — never wandb directly.
"""

import wandb
from pathlib import Path


class ExperimentLogger:
    """
    Thin wrapper around W&B. Instantiate once at the start of a training run.

    Usage:
        logger = ExperimentLogger(config)
        logger.log_epoch(epoch=1, train_metrics={...}, val_metrics={...})
        logger.log_model(checkpoint_path="checkpoints/best.pt", metric_value=6.4)
        logger.finish()
    """

    def __init__(self, config: dict, run_name: str | None = None, tags: list[str] | None = None):
        """
        Args:
            config:   Full config dict (loaded from default.yaml). Logged as hyperparameters.
            run_name: Optional human-readable name. Auto-generated by W&B if None.
            tags:     Optional list of tags, e.g. ["sweep", "ablation-spatial-only"].
        """
        self.run = wandb.init(
            project="airsketch",
            entity="airsketch-umd",
            name=run_name,
            config=config,
            tags=tags or [],
            resume="allow",
        )
        self.run_name = self.run.name
        self.run_id   = self.run.id
        print(f"W&B run started: {self.run.url}")

    # ── Per-epoch logging ─────────────────────────────────────────────────────

    def log_epoch(
        self,
        epoch: int,
        train_metrics: dict,
        val_metrics: dict,
        lr: float,
    ) -> None:
        """
        Log all metrics for one epoch.

        Expected keys in train_metrics / val_metrics:
            mpjpe        (float) — mean per-joint position error in pixels
            jitter_index (float) — mean frame-to-frame displacement variance px²
            gesture_acc  (float) — draw/idle classification accuracy [0, 1]
            loss         (float) — combined joint loss value

        Args:
            epoch:         Current epoch number (1-indexed).
            train_metrics: Dict of training metrics for this epoch.
            val_metrics:   Dict of validation metrics for this epoch.
            lr:            Current learning rate (from scheduler).
        """
        payload = {"epoch": epoch, "lr": lr}

        for key, value in train_metrics.items():
            payload[f"train/{key}"] = value

        for key, value in val_metrics.items():
            payload[f"val/{key}"] = value

        wandb.log(payload, step=epoch)

    # ── Step-level logging (optional, for loss curves within an epoch) ────────

    def log_step(self, step: int, loss: float, gesture_loss: float, regression_loss: float) -> None:
        """Log per-batch losses. Call every N steps (e.g. every 10 batches)."""
        wandb.log({
            "step/loss":            loss,
            "step/gesture_loss":    gesture_loss,
            "step/regression_loss": regression_loss,
        }, step=step)

    # ── Model checkpointing ───────────────────────────────────────────────────

    def log_model(self, checkpoint_path: str | Path, metric_value: float, metric_name: str = "val/mpjpe") -> None:
        """
        Save a model checkpoint as a W&B artifact.
        Call this whenever a new best checkpoint is saved.

        Args:
            checkpoint_path: Path to the .pt file on disk.
            metric_value:    The metric value that triggered this checkpoint save.
            metric_name:     Name of the metric used for best-model selection.
        """
        artifact = wandb.Artifact(
            name=f"airsketch-model-{self.run_id}",
            type="model",
            metadata={metric_name: metric_value},
        )
        artifact.add_file(str(checkpoint_path))
        self.run.log_artifact(artifact)
        print(f"Checkpoint logged to W&B artifact: {checkpoint_path} ({metric_name}={metric_value:.4f})")

    # ── Summary metrics (end of run) ──────────────────────────────────────────

    def log_summary(self, test_metrics: dict) -> None:
        """
        Log final test-set metrics at the end of a run.
        These appear in the W&B run summary panel.

        Args:
            test_metrics: Dict with keys mpjpe, jitter_index, gesture_acc, latency_ms, sus_score.
        """
        for key, value in test_metrics.items():
            wandb.run.summary[f"test/{key}"] = value

    # ── Cleanup ───────────────────────────────────────────────────────────────

    def finish(self) -> None:
        """Call at the end of every training script. Flushes remaining logs."""
        wandb.finish()
        print(f"W&B run finished. View at: https://wandb.ai/airsketch-umd/airsketch/runs/{self.run_id}")
```

---

### 3. Metric Definitions

These are the exact keys the logger expects. Define them once here so every team member computes them identically. Add this to `src/evaluate.py` (stub file from issue #1):

```python
"""
src/evaluate.py
Metric computation functions for AirSketch.
All functions return Python floats. Units are documented in the docstring.
"""

import numpy as np


def compute_mpjpe(pred_xy: np.ndarray, gt_xy: np.ndarray) -> float:
    """
    Mean Per-Joint Position Error — average Euclidean distance between
    predicted and ground-truth index fingertip coordinates in pixels.

    Target: < 8 px on 224×224 frames.

    Args:
        pred_xy: (N, 2) array of predicted (x, y) coordinates.
        gt_xy:   (N, 2) array of ground-truth (x, y) coordinates.

    Returns:
        MPJPE in pixels (float).
    """
    assert pred_xy.shape == gt_xy.shape, "Shape mismatch between predictions and ground truth"
    return float(np.mean(np.linalg.norm(pred_xy - gt_xy, axis=1)))


def compute_jitter_index(pred_xy: np.ndarray) -> float:
    """
    Stroke Smoothness — mean frame-to-frame displacement variance of
    predicted fingertip positions. Lower = smoother strokes.

    Target: < 2 px² variance.

    Args:
        pred_xy: (N, 2) array of predicted (x, y) coordinates in sequence order.

    Returns:
        Jitter index in px² (float).
    """
    if len(pred_xy) < 2:
        return 0.0
    frame_deltas = np.linalg.norm(np.diff(pred_xy, axis=0), axis=1)
    return float(np.var(frame_deltas))


def compute_gesture_accuracy(pred_labels: np.ndarray, gt_labels: np.ndarray) -> float:
    """
    Binary classification accuracy for draw vs idle gesture.

    Target: > 0.92 (92%).

    Args:
        pred_labels: (N,) array of predicted labels {0: idle, 1: draw}.
        gt_labels:   (N,) array of ground-truth labels {0: idle, 1: draw}.

    Returns:
        Accuracy as a float in [0, 1].
    """
    assert pred_labels.shape == gt_labels.shape
    return float(np.mean(pred_labels == gt_labels))


def compute_all_metrics(pred_xy, gt_xy, pred_labels, gt_labels) -> dict:
    """
    Compute all three tracked metrics in one call.
    Returns a dict ready to pass directly to ExperimentLogger.log_epoch().
    """
    return {
        "mpjpe":        compute_mpjpe(pred_xy, gt_xy),
        "jitter_index": compute_jitter_index(pred_xy),
        "gesture_acc":  compute_gesture_accuracy(pred_labels, gt_labels),
    }
```

---

### 4. W&B Sweep Config for Issue #11

Create this file now so issue #11 can use it directly without any setup work. Commit it to `configs/sweep.yaml`:

```yaml
# configs/sweep.yaml
# W&B hyperparameter sweep configuration for AirSketch.
# Issue #11 runs this sweep. Issue #3 commits the file.
#
# Launch with:
#   wandb sweep configs/sweep.yaml          → prints a sweep ID
#   wandb agent airsketch-umd/airsketch/<sweep-id>   → starts a worker

program: src/train.py
method: grid

metric:
  name: val/mpjpe
  goal: minimize

early_terminate:
  type: hyperband
  min_iter: 5
  eta: 2

parameters:
  model.hidden_dim:
    values: [64, 128, 256]

  model.num_layers:
    values: [2, 4, 6]

  training.learning_rate:
    values: [0.001, 0.0005, 0.0001]

  # Fixed for the sweep — not being searched
  training.batch_size:
    value: 64

  training.epochs:
    value: 50

  model.dropout:
    value: 0.1
```

To launch the sweep from Zaratan:

```bash
# Step 1 — register the sweep (run once from login node)
wandb sweep configs/sweep.yaml
# Output: wandb: Created sweep with ID: <sweep-id>
#         wandb: View sweep at: https://wandb.ai/airsketch-umd/airsketch/sweeps/<sweep-id>

# Step 2 — submit one SLURM job per sweep agent (each agent picks up one config at a time)
# Edit scripts/slurm/sweep.sh to replace the array approach with:
#   wandb agent airsketch-umd/airsketch/<sweep-id>
sbatch scripts/slurm/sweep.sh
```

---

### 5. Integration Example — How Training Code Uses the Logger

Show exactly how issue #10's training loop should call the logger. This goes in `src/train.py`:

```python
from src.logger import ExperimentLogger
from src.evaluate import compute_all_metrics
import yaml

def train(config_path: str, run_name: str | None = None):
    with open(config_path) as f:
        config = yaml.safe_load(f)

    logger = ExperimentLogger(config=config, run_name=run_name, tags=["training"])

    for epoch in range(1, config["training"]["epochs"] + 1):

        # --- training pass (pseudocode) ---
        train_preds_xy, train_gt_xy       = run_epoch(split="train")
        train_preds_lbl, train_gt_lbl     = run_gesture_epoch(split="train")
        train_loss                         = compute_loss(...)

        # --- validation pass ---
        val_preds_xy, val_gt_xy           = run_epoch(split="val")
        val_preds_lbl, val_gt_lbl         = run_gesture_epoch(split="val")
        val_loss                           = compute_loss(...)

        # --- compute metrics ---
        train_metrics = compute_all_metrics(train_preds_xy, train_gt_xy, train_preds_lbl, train_gt_lbl)
        val_metrics   = compute_all_metrics(val_preds_xy,   val_gt_xy,   val_preds_lbl,   val_gt_lbl)

        train_metrics["loss"] = train_loss
        val_metrics["loss"]   = val_loss

        # --- log everything ---
        logger.log_epoch(
            epoch=epoch,
            train_metrics=train_metrics,
            val_metrics=val_metrics,
            lr=scheduler.get_last_lr()[0],
        )

        # --- checkpoint if best val MPJPE ---
        if val_metrics["mpjpe"] < best_mpjpe:
            best_mpjpe = val_metrics["mpjpe"]
            save_checkpoint(model, path="checkpoints/best.pt")
            logger.log_model("checkpoints/best.pt", metric_value=best_mpjpe)

    logger.finish()
```

---

### 6. Viewing Runs — Reference Notebook

Create `notebooks/04_results_visualization.ipynb` with the following cells. Commit it to the repo with pre-run output cleared (so it doesn't inflate the repo size):

**Cell 1 — Install and authenticate**
```python
import wandb
wandb.login()
```

**Cell 2 — Load all runs from the project**
```python
api = wandb.Api()
runs = api.runs("airsketch-umd/airsketch")

import pandas as pd

records = []
for run in runs:
    if run.state != "finished":
        continue
    records.append({
        "name":          run.name,
        "hidden_dim":    run.config.get("model", {}).get("hidden_dim"),
        "num_layers":    run.config.get("model", {}).get("num_layers"),
        "lr":            run.config.get("training", {}).get("learning_rate"),
        "val_mpjpe":     run.summary.get("val/mpjpe"),
        "val_jitter":    run.summary.get("val/jitter_index"),
        "val_gesture":   run.summary.get("val/gesture_acc"),
        "best_epoch":    run.summary.get("epoch"),
    })

df = pd.DataFrame(records).sort_values("val_mpjpe")
df
```

**Cell 3 — Plot MPJPE over epochs for all runs**
```python
import matplotlib.pyplot as plt

fig, axes = plt.subplots(1, 3, figsize=(15, 4))
metrics = ["val/mpjpe", "val/jitter_index", "val/gesture_acc"]
titles  = ["Val MPJPE (px) ↓", "Val Jitter Index (px²) ↓", "Val Gesture Acc ↑"]
targets = [8.0, 2.0, 0.92]

for ax, metric, title, target in zip(axes, metrics, titles, targets):
    for run in runs:
        if run.state != "finished":
            continue
        history = run.history(keys=[metric])
        if metric in history.columns:
            ax.plot(history[metric].values, alpha=0.6, label=run.name)
    ax.axhline(y=target, color="red", linestyle="--", linewidth=1.5, label=f"Target: {target}")
    ax.set_title(title)
    ax.set_xlabel("Epoch")
    ax.legend(fontsize=7, loc="best")

plt.tight_layout()
plt.savefig("report/figures/training_curves.png", dpi=150, bbox_inches="tight")
plt.show()
```

**Cell 4 — Sweep results heatmap (run after issue #11)**
```python
import seaborn as sns
import numpy as np

pivot = df.pivot_table(
    index="num_layers",
    columns="hidden_dim",
    values="val_mpjpe",
    aggfunc="min"
)

fig, ax = plt.subplots(figsize=(7, 4))
sns.heatmap(
    pivot,
    annot=True,
    fmt=".2f",
    cmap="YlOrRd_r",
    linewidths=0.5,
    ax=ax,
    cbar_kws={"label": "Val MPJPE (px)"}
)
ax.set_title("Sweep results — val MPJPE by hidden_dim × num_layers\n(lower is better, target < 8 px)")
ax.set_xlabel("Hidden dim")
ax.set_ylabel("Num layers")
plt.tight_layout()
plt.savefig("report/figures/sweep_heatmap.png", dpi=150, bbox_inches="tight")
plt.show()
```

**Cell 5 — Print best run config**
```python
best = df.iloc[0]
print("Best run:")
print(f"  Name:        {best['name']}")
print(f"  hidden_dim:  {best['hidden_dim']}")
print(f"  num_layers:  {best['num_layers']}")
print(f"  lr:          {best['lr']}")
print(f"  val MPJPE:   {best['val_mpjpe']:.3f} px  (target < 8 px)")
print(f"  val jitter:  {best['val_jitter']:.3f} px² (target < 2 px²)")
print(f"  val gesture: {best['val_gesture']:.3f}    (target > 0.92)")
```

---

### 7. Checklist

- [ ] W&B team `airsketch-umd` created and all four members added
- [ ] API key shared securely (use a private Slack DM or UMD email — never commit it)
- [ ] `wandb login` verified on each team member's local machine
- [ ] `wandb login` verified on Zaratan inside the conda environment
- [ ] `src/logger.py` committed with `ExperimentLogger` class
- [ ] `src/evaluate.py` updated with `compute_mpjpe`, `compute_jitter_index`, `compute_gesture_accuracy`, `compute_all_metrics`
- [ ] `configs/sweep.yaml` committed
- [ ] `notebooks/04_results_visualization.ipynb` committed with output cleared
- [ ] Offline mode tested on Zaratan — `WANDB_MODE=offline` + `wandb sync` workflow confirmed working
- [ ] A smoke-test run submitted (even with a dummy model) — confirm metrics appear in the W&B dashboard
- [ ] Link to the W&B project dashboard added to `README.md`

---

### Definition of done

After a training run completes on Zaratan, any team member can open the W&B project at `https://wandb.ai/airsketch-umd/airsketch`, see the run listed with correct hyperparameters, and view epoch-by-epoch curves for `val/mpjpe`, `val/jitter_index`, and `val/gesture_acc` without any manual export or copy-paste.

**Comment by sachelsout** [comment]
Closing this out — everything on the checklist is done.

**What was implemented:**
- `src/logger.py` — `ExperimentLogger` wrapper around W&B. All training code should call this, never `wandb` directly.
- `src/evaluate.py` — `compute_mpjpe`, `compute_jitter_index`, `compute_gesture_accuracy`, `compute_all_metrics` with targets documented in docstrings.
- `configs/sweep.yaml` — grid search over `hidden_dim × num_layers × lr` (27 configs), ready for issue #11.
- `notebooks/04_results_visualization.ipynb` — training curves, sweep heatmap, best-run summary.
- README updated with W&B dashboard link.

**Zaratan setup notes:**
- `wandb login` doesn't work on Zaratan due to a `gevent` binary incompatibility in the cvmfs stack. Workaround: set `WANDB_API_KEY` as an environment variable in `~/.bashrc` and all SLURM scripts.
- Offline mode tested and confirmed working (`WANDB_MODE=offline` + `wandb sync`). Use this as fallback if outbound traffic is ever blocked during a job.
- Smoke test run synced successfully — metrics visible at https://wandb.ai/airsketch/airsketch

**For teammates:** Add `export WANDB_API_KEY=<key>` to your `~/.bashrc` on Zaratan and your local machine. Key is in our WhatsApp group. Entity is `airsketch`, not `airsketch-umd`.

Blocks #10 and #11 are now unblocked.
